zcash_primitives: Add explicit control over padding to the transaction builder. (#1065)
zcash/librustzcash: zcash_primitives/CHANGELOG.md

- Dependency on `bellman 0.14`.
- `zcash_primitives::consensus::sapling_zip212_enforcement`
- `zcash_primitives::transaction`:

>   - `builder::{get_fee, PaddingRules}`

Reviewer: ```suggestion
  - `builder::{BuildConfig, FeeError, get_fee}`
```

  - `builder::Builder` now has a generic parameter for the type of progress
    notifier, which needs to implement `sapling::builder::ProverProgress` in
    order to build transactions.

>   - `builder::Builder::{new, new_with_rng}` now take an additional `PaddingRules`
>     argument.

Reviewer: `Builder::new_with_rng` has been removed.
```suggestion
  - `builder::Builder::new` now take a `BuildConfig` argument instead of an
    optional Orchard anchor.
```

  - `builder::Builder` now has a generic parameter for the type of progress
    notifier, which needs to implement `sapling::builder::ProverProgress` in
    order to build transactions.
  - `builder::Builder::new` now takes a `BuildConfig` argument instead of an
    optional Orchard anchor. Anchors for both Sapling and Orchard are now 
    required at the time of builder construction.
  - `builder::Builder::{build, build_zfuture}` now take
    `&impl SpendProver, &impl OutputProver` instead of `&impl TxProver`.
  - `builder::Builder::add_sapling_spend` no longer takes a `diversifier`
    argument as the diversifier may be obtained from the note.
  - `builder::Builder::add_sapling_spend` now takes its `ExtendedSpendingKey` 
    argument by reference.

>   - `builder::Builder::{add_sapling_spend, add_sapling_output}` now return a 

Reviewer: ```suggestion
  - `builder::Builder::{add_sapling_spend, add_sapling_output}` now return
```

    `builder::Error`s instead of the underlying `sapling_crypto::builder::Error`s 
    when returning `Err`.
  - `builder::Builder::add_orchard_spend` now takes its `SpendingKey` argument
     by reference.
  - `builder::Builder::with_progress_notifier` now consumes `self` and returns a
    `Builder` typed on the provided channel.
  - `builder::Builder::get_fee` now returns a `builder::FeeError` instead of the
    bare `FeeRule::Error` when returning `Err`.
  - `builder::Error::OrchardAnchorNotAvailable` has been renamed to 

>     `OrchardBuilderNotAvailable` has been renamed to 

Reviewer: ```suggestion
    `OrchardBuilderNotAvailable`.
```

    - `Bundle`
    - `SpendDescription, SpendDescriptionV5`
    - `OutputDescription, OutputDescriptionV5`

>     - `Authorization, Authorized, MapAuth`

Reviewer: The changes that were made to `MapAuth` in `sapling-crypto` need to be moved back into this changelog, because the `MapAuth` trait added back by this PR does not match the `MapAuth` trait in `zcash_primitives 0.13.0`. I think the only change is `&self` being changed to `&mut self`, but please check this and add the corresponding change entry.
